### DISCUSSION TITLE
Computing Earth Movers Distance

### DISCUSSION THREAD

**Discussion by jamied157** [synthesized-io/insight#23] [General]
So currently we have two ways of computing EMD, the first is found here which essentially assumes each category is equidistant from one another:
https://github.com/synthesized-io/insight/blob/8d17059f3f960f0f77c85e96e581115867c7b39d/src/synthesized_insight/metrics/metrics.py#L197-L241

The second is here which I think takes an already binned set of variables and uses the centre of the bins to calculate the distance:
https://github.com/synthesized-io/insight/blob/f88cf964feb685d0265ba07f2f4064e61af5cbb1/src/synthesized_insight/metrics/distance.py#L358-L405

I think we should have a way of calculating distances for affine variables even if they aren't binned. We should check these metrics agree with pyemd.

**Comment by simonhkswan** [comment]
We have this in the SDK, so shall we move it across?

https://github.com/synthesized-io/synthesized/blob/a894520b7e61abba7b06a94a22377ea598ac6bfc/synthesized/insight/metrics/distance.py#L485-L498

```python
class EarthMoversDistance(DistanceMetric):
    """
    Earth movers distance (EMD), aka Wasserstein 1-distance, between samples from two distributions.
    """

    @property
    def distance(self) -> float:
        """
        Calculate the EMD between two 1d histograms.
        Returns:
            The earth mover's distance.
        """
        return wasserstein_distance(self.x, self.y)
```

**Comment by simonhkswan** [comment]
As for comparing to pyemd, this was done when the code was swapped: synthesized-io/synthesized#564

**Reply by jamied157** [reply]
Great thanks for sharing!
